Sources/SwiftDocC/Infrastructure/Link Resolution/ExternalPathHierarchyResolver.swift

```diff
@@ -0,0 +1,235 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2023 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import SymbolKit
+import Markdown
+
+/// A class that resolves links to an already
+final class ExternalPathHierarchyResolver {
+    /// A hierarchy of path components used to resolve links in the documentation.
+    private(set) var pathHierarchy: PathHierarchy!
+    
+    /// A map from the path hierarchies identifiers to resolved references.
+    private var resolvedReference = [ResolvedIdentifier: ResolvedTopicReference]()
+    /// A map from symbol's unique identifiers to their resolved references.
+    private var symbols: [String: ResolvedTopicReference]
+    
+    /// The content for each external entity.
+    private var content: [ResolvedTopicReference: LinkDestinationSummary]
+    
+    /// Attempts to resolve an unresolved reference.
+    ///
+    /// - Parameters:
+    ///   - unresolvedReference: The unresolved reference to resolve.
+    ///   - isCurrentlyResolvingSymbolLink: Whether or not the documentation link is a symbol link.
+    /// - Returns: The result of resolving the reference.
+    func resolve(_ unresolvedReference: UnresolvedTopicReference, fromSymbolLink isCurrentlyResolvingSymbolLink: Bool) -> TopicReferenceResolutionResult {
+        do {
+            let foundID = try pathHierarchy.find(path: Self.path(for: unresolvedReference), parent: nil, onlyFindSymbols: true)
+            guard let foundReference = resolvedReference[foundID] else {
+                fatalError("Every identifier in the path hierarchy has a corresponding reference in the wrapping resolver. If it doesn't that's an indication that the file content that it was deserialized from was malformed.")
+            }
+            
+            guard content[foundReference] != nil else {
+                return .failure(unresolvedReference, .init("Resolved \(foundReference.url.withoutHostAndPortAndScheme().absoluteString.singleQuoted) but don't have any content to display for it."))
+            }
+            
+            return .success(foundReference)
+        } catch let error as PathHierarchy.Error {
+            var originalReferenceString = unresolvedReference.path
+            if let fragment = unresolvedReference.topicURL.components.fragment {
+                originalReferenceString += "#" + fragment
```

Comment:
I changed this to use `Self.path` to only compute the original string once.

I would prefer to wait for one or two more places that need something like this so that we can have a better sense of what's needed for an API like this.